source/extensions/filters/http/grpc_stats/grpc_stats_filter.cc

```diff
@@ -17,66 +19,186 @@ namespace GrpcStats {
 
 namespace {
 
+// A map from gRPC service/method name to symbolized stat names for the service/method.
+//
+// The expected usage pattern is that the map is populated once, and can then be queried lock-free
+// as long as it isn't being modified.
+class GrpcServiceMethodToRequestNamesMap {
+public:
+public:
+  // Construct a map populated with the services/methods in method_list.
+  GrpcServiceMethodToRequestNamesMap(Stats::SymbolTable& symbol_table,
+                                     const envoy::config::core::v3::GrpcMethodList& method_list)
+      : stat_name_pool_(symbol_table), map_(populate(method_list)) {}
+
+  absl::optional<Grpc::Context::RequestStatNames>
+  lookup(const Grpc::Common::RequestNames& request_names) const {
+    auto it = map_.find(ViewKey(request_names.service_, request_names.method_));
+    if (it != map_.end()) {
+      return it->second;
+    }
+
+    return {};
+  }
+
+private:
+  using OwningKey = std::tuple<std::string, std::string>;
+  using ViewKey = std::tuple<absl::string_view, absl::string_view>;
+  struct MapHash : absl::Hash<OwningKey>, absl::Hash<ViewKey> {
+    using is_transparent = void;
+  };
+
+  struct MapEq {
+    using is_transparent = void;
+    bool operator()(const OwningKey& left, const OwningKey& right) const { return left == right; }
+    bool operator()(const OwningKey& left, const ViewKey& right) const { return left == right; }
+  };
+  using MapType = absl::flat_hash_map<OwningKey, Grpc::Context::RequestStatNames, MapHash, MapEq>;
+
+  // Helper for generating a populated MapType so that `map_` can be const.
+  MapType populate(const envoy::config::core::v3::GrpcMethodList& method_list) {
+    MapType map;
+    for (const auto& service : method_list.services()) {
+      Stats::StatName stat_name_service = stat_name_pool_.add(service.name());
+
+      for (const auto& method_name : service.method_names()) {
+        Stats::StatName stat_name_method = stat_name_pool_.add(method_name);
+        map[OwningKey(service.name(), method_name)] =
```

Comment:
Your gcc build failed here with
```
external/com_google_absl/absl/container/internal/raw_hash_set.h:1324:32: error: request for member 'operator()' is ambiguous
     return find(key, hash_ref()(key));
```

One thing you did slightly different than I expected is you made MapHash multiply inherit from the two hashers. That form makes sense to me, but maybe you could try:
```
  struct MapHash {
    using is_transparent = void;
    uint64_t operator()(const OwningKey& key) const { return absl::Hash<OwningKey>()(key); }
    uint64_t operator()(const ViewKey& key) const { return absl::Hash<ViewKey>()(key); }
  };
```
If you can set up local iteration with gcc maybe you can figure out what to tweak.

Reply:
Yeah, I saw the failure right at the end of the day yesterday. I'll sort this out today. I think the problem, from gcc's view, is that OwningKey can be constructed from ViewKey, so both hash options are possible, because std::tuple has some fancy stuff. I may try to directly use Grpc::Common::StatsNames as ViewKey instead of a std::tuple. I'll figure something out.